x/lockup/keeper/lock.go

```diff
@@ -560,3 +566,43 @@ func (k Keeper) Unlock(ctx sdk.Context, lock types.PeriodLock) error {
 	k.hooks.OnTokenUnlocked(ctx, owner, lock.ID, lock.Coins, lock.Duration, lock.EndTime)
 	return nil
 }
+
+// ForceUnlock ignores unlock duration and immediately unlock and refund.
+// CONTRACT: should be used only at the chain upgrade script
+func (k Keeper) ForceUnlock(ctx sdk.Context, lock types.PeriodLock) error {
+	owner, err := sdk.AccAddressFromBech32(lock.Owner)
+	if err != nil {
+		return err
+	}
+
+	// send coins back to owner
+	if err := k.bk.SendCoinsFromModuleToAccount(ctx, types.ModuleName, owner, lock.Coins); err != nil {
+		return err
+	}
+
+	// remove lock from store object
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(lockStoreKey(lock.ID))
+
+	var keyprefix []byte
+	if lock.IsUnlocking() {
+		keyprefix = types.KeyPrefixUnlocking
+	} else {
+		keyprefix = types.KeyPrefixNotUnlocking
+	}
+
+	// delete lock refs from the unlocking queue
+	err = k.deleteLockRefs(ctx, keyprefix, lock)
+	if err != nil {
+		return err
+	}
+
+	// remove from accumulation store
+	for _, coin := range lock.Coins {
+		k.accumulationStore(ctx, coin.Denom).Decrease(accumulationKey(lock.Duration), coin.Amount)
+	}
+
+	k.hooks.OnTokenUnlocked(ctx, owner, lock.ID, lock.Coins, lock.Duration, lock.EndTime)
+	return nil
```

Comment:
Can we reuse logic from the existing methods for a lock going from NotUnlocking -> Unlocking, and Unlocking -> finish unlocking?

Reply:
Will do so!

Reply:
Problem is that the existing method checks whether the time has been passed specified on the lock duration, and ForceUnlock need to ignore them.

Reply:
Fixed by separating duration validation and actual unlocking logic